gui/liquids.lua

```diff
@@ -1,336 +1,195 @@
-LiquidsUI = defclass(LiquidsUI, guidm.MenuOverlay)
-
-LiquidsUI.ATTRS = {
-    sidebar_mode=df.ui_sidebar_mode.LookAround,
-}
-
-LiquidsUI.focus_path = 'liquids'
-
-function LiquidsUI:init()
-    self:assign{
-        brush = Toggle{ items = brushes },
-        paint = Toggle{ items = paints },
-        flow = Toggle{ items = flowbits },
-        set = Toggle{ items = setmode },
-        permaflow = Toggle{ items = permaflows },
-        amount = 7,
-    }
-end
-
-function LiquidsUI:onDestroy()
-    guidm.clearSelection()
-end
-
-function render_liquid(dc, block, x, y)
-    local dsgn = block.designation[x%16][y%16]
-
-    if dsgn.flow_size > 0 then
-        if dsgn.liquid_type == df.tile_liquid.Magma then
-            dc:pen(COLOR_RED):string("Magma")
-        else
-            dc:pen(COLOR_BLUE)
-            if dsgn.water_stagnant then dc:string("Stagnant ") end
-            if dsgn.water_salt then dc:string("Salty ") end
-            dc:string("Water")
-        end
-        dc:string(" ["..dsgn.flow_size.."/7]")
-    else
-        dc:string('No Liquid')
+function SpawnLiquid:decreaseLiquidLevel()
+    if self.level > 1 then
+        self.level = self.level - 1
     end
 end
 
-local permaflow_abbr = {
-    north = 'N', south = 'S', east = 'E', west = 'W',
-    northeast = 'NE', northwest = 'NW', southeast = 'SE', southwest = 'SW'
-}
-
-function render_flow_state(dc, block, x, y)
-    local flow = block.liquid_flow[x%16][y%16]
-
-    if block.flags.update_liquid then
-        dc:string("Updating", COLOR_GREEN)
-    else
-        dc:string("Static")
-    end
-    dc:string(", ")
-    if flow.perm_flow_dir ~= 0 then
-        local tag = df.tile_liquid_flow_dir[flow.perm_flow_dir]
-        dc:string("Permaflow "..(permaflow_abbr[tag] or tag), COLOR_CYAN)
-    elseif flow.temp_flow_timer > 0 then
-        dc:string("Flowing "..flow.temp_flow_timer, COLOR_GREEN)
-    else
-        dc:string("No Flow")
+function SpawnLiquid:spawn(pos)
+    if dfhack.maps.isValidTilePos(pos) and dfhack.maps.isTileVisible(pos) then
+        liquids.spawnLiquid(pos, self.level, self.type)
     end
 end
 
-function LiquidsUI:onRenderBody(dc)
-    dc:clear():seek(1,1):string("Paint Liquids Cheat", COLOR_WHITE)
+function SpawnLiquid:getPen()
+    local tile = nil
 
-    local cursor = guidm.getCursorPos()
-    local block = dfhack.maps.getTileBlock(cursor)
-
-    if block then
-        local x, y = pos2xyz(cursor)
-        local tile = block.tiletype[x%16][y%16]
-
-        dc:seek(2,3):string(df.tiletype.attrs[tile].caption, COLOR_CYAN)
-        dc:newline(2):pen(COLOR_DARKGREY)
-        render_liquid(dc, block, x, y)
-        dc:newline(2):pen(COLOR_DARKGREY)
-        render_flow_state(dc, block, x, y)
+    if self.type == df.tile_liquid.Water then
+        tile = dfhack.screen.findGraphicsTile('MINING_INDICATORS', 0, 0)
     else
-        dc:seek(2,3):string("No map data", COLOR_RED):advance(0,2)
+        tile = dfhack.screen.findGraphicsTile('MINING_INDICATORS', 1, 0)
```

Comment:
you can do the tile lookups once in the top scope (above this function declaration) instead of once on every tile. the `findGraphicsTile()` function is linear in the number of loaded png files and is not exactly fast.